enhancements/migration/cam-non-admin-design.md

```diff
@@ -0,0 +1,351 @@
+work. I will break down the required changes for each component:
+
+#### Migration Operator
+
+For the current design, there will exist one operator running in
+`openshift-migration`. For an alternative approach, please see OLM
+OperatorGroup Integration below.
+
+There will be at least one new CRD introduced which will define a namespaced
+installation of CAM. The operator will be watching ALL namespaces for the
+creation of this CR. For now, this will be called a `MigrationSandbox`. In the
+current design, when a MigrationSandbox is created in namespace `Foo`, the
+operator will install a namespaced Migration UI in namespace `Foo`. A new
+Migration Controller will be deployed in `openshift-migration` which will be
+watching namespace `Foo` for the creation of Migration Resources.
+
+#### Migration Controller
+
+The migration controller needs to run in `openshift-migration` since it will
+contain access to the Mig SA token. We must not allow the user to be able to
+exec into the controller pod and escalate their privileges.
+
+The controller’s migplan controller will need to be updated to validate that
+the list of namespaces selected on the migplan do not include namespaces which
+a user does not have access to migrate. To do this, it will leverage the
+discovery service’s RBAC-based authorization support to determine what
+namespaces the user can migrate.
+
+##### Discovery Controller
+
+The discovery controller will be updated to support RBAC-based authorization.
+Since CAM will be updated to allow a user to specify an identity token on the
+migration plan, the discovery controller can take this identity token in and
+use it as a bearer token.
+
+The discovery controller would inventory the kubernetes RBAC resources and
+effectively implement Subject Access Review using the bearer token.  The
+authentication would be delegated to the cluster using a TokenReview. The RBAC
+model in kubernetes seems to be straight-forward and consistent with commonly
+found models. Based on the quality of the model and documentation, the
+implementation of the Access Review seems to be equally straight-forward.  This
+isn’t re-inventing RBAC, it’s simply applying the kubernetes RBAC rule set to
+the discovered data to provide authorization. Mainly leveraging an existing
+authentication mechanism and set of RBAC rules already being managed by users.
+
+The controller inventories (and watches):
+* ClusterRoleBinding
+* RoleBinding
+* ClusterRole
+* Role
+
+The data is collected using the Mig SA token associated with the
+MigCluster.  The data is then indexed and used in conjunction with the user's
+identity token to determine the list of resources the user is able to migrate.
+
+REST request authorization steps:
+1. Authenticate the Bearer token to a User or ServiceAccount by performing a
+   TokenReview.  This step will provide information about the token (user or
+   SA).
+1. Find ClusterRoleBindings/RoleBindings for the User (and groups) or
```

Comment:
I am concerned about writing the authorization code. I understand that the model is well documented, and it seems simple but it seems like something we should not be attempting. 

> https://kubernetes.io/docs/reference/access-authn-authz/authorization/#using-flags-for-your-authorization-module

> You can choose more than one authorization module. Modules are checked in order so an earlier module has higher priority to allow or deny a request.

For pure kubernetes or other kubernetes sources, this is where I see the problem being. It would not be safe to assume that `user A` has access just because of RBAC. 

If you are on openshift by itself, it seems more prudent to just use a subject rules review, and relay on well-tested default authorization logic IMO.

Reply:
We agree on the limitation that with pure Kubernetes RBAC might not be sufficient for determining what resources a user actually has access to. We are limiting the scope of this to *only* support limiting access based on RBAC and we are scoping the initial design to support OCP->OCP as the primary use case.

While we thought SRR was going to be a viable solution, when we dug into it we realized that  with the number of reconciles our controllers are running it would require us to make hundreds/thousands of SRR a minute. From the design document where we rejected it, here are the expected number of API requests if we used SRR:

```
10 namespaces  = 660 cluster API requests / minute.  (11 / second)
30 namespaces = 1860 cluster API requests / minute. (31 / second)
60 namespaces = 3660 cluster API requests / minute. (61 / second)
90 namespaces = 5460 cluster API requests / minute. (91 / second)
120 namespaces = 7260 cluster API requests / minute. (121 / second)
```

This option also left us with no room to grow into pure Kubernetes clusters as well.

I will add a blurb about why we rejected this option in the doc.

Reply:
Can you explain the numbers? 

10 namespaces = 660 cluster API requests?

Reply:
Sure, from the design doc:

>As a reminder, the Plan controller reconciles between 1 / 2-3 seconds and 2-3 / second. For the purposes of discussion, let’s use an average of 60 SRR requests / minute.  Using an average of 10 namespaces on a migration plan, this equates to 600 SRR requests / minute.  Plus 60 project list requests = 660 cluster API requests / minute (sustained rate).  This is with only 10 namespaces and only includes the Plan the namespace validation use case.  Also, this is based on observations made on my development clusters with no workloads.  It’s reasonable to expect that a customer production cluster with thousands of workloads running will result in a multiple of this estimated rate of Plan reconcile.

If you'd like to read more about why we rejected the SRR approach please see the [design document](https://docs.google.com/document/d/1Ic5bfmwtMM_AV843neQ7KNKPfgJD7-ADgreMcZaMo4E/edit) under *Option B*